apps/cowswap-frontend/src/common/hooks/useContract.ts

```diff
@@ -12,73 +12,106 @@ import {
   Weth,
   WethAbi,
 } from '@cowprotocol/abis'
-import { V_COW_CONTRACT_ADDRESS, WRAPPED_NATIVE_CURRENCIES } from '@cowprotocol/common-const'
-import { getContract, getEthFlowContractAddress } from '@cowprotocol/common-utils'
-import { COW_PROTOCOL_SETTLEMENT_CONTRACT_ADDRESS } from '@cowprotocol/cow-sdk'
+import {
+  COWSWAP_ETHFLOW_CONTRACT_ADDRESS,
+  V_COW_CONTRACT_ADDRESS,
+  WRAPPED_NATIVE_CURRENCIES,
+} from '@cowprotocol/common-const'
+import { getContract, isEns, isProd, isStaging } from '@cowprotocol/common-utils'
+import { COW_PROTOCOL_SETTLEMENT_CONTRACT_ADDRESS, SupportedChainId } from '@cowprotocol/cow-sdk'
 import { useWalletInfo } from '@cowprotocol/wallet'
 import { useWalletProvider } from '@cowprotocol/wallet-provider'
 import { Contract } from '@ethersproject/contracts'
 import { Web3Provider } from '@ethersproject/providers'
 
+const WETH_CONTRACT_ADDRESS_MAP = Object.fromEntries(
+  Object.entries(WRAPPED_NATIVE_CURRENCIES).map(([chainId, token]) => [chainId, token.address]),
+)
+
+const contractEnv = isProd || isStaging || isEns ? 'prod' : 'barn'
+export const COWSWAP_ETHFLOW_CONTRACT_ADDRESS_MAP = COWSWAP_ETHFLOW_CONTRACT_ADDRESS[contractEnv]
+
+export type UseContractResult<T extends Contract = Contract> = {
+  contract: T | null
+  error: unknown | null
+  loading: boolean
+  chainId: SupportedChainId
+}
+
 // returns null on errors
 export function useContract<T extends Contract = Contract>(
-  addressOrAddressMap: string | { [chainId: number]: string } | undefined,
+  addressOrAddressMap: string | { [chainId: number]: string | undefined | null } | undefined,
   ABI: any,
   withSignerIfPossible = true,
-  customProvider?: Web3Provider
-): T | null {
+  customProvider?: Web3Provider,
+): UseContractResult<T> {
   const defaultProvider = useWalletProvider()
   const { account, chainId } = useWalletInfo()
   const provider = customProvider || defaultProvider
 
   return useMemo(() => {
-    if (!addressOrAddressMap || !ABI || !provider || !chainId) return null
-    let address: string | undefined
-    if (typeof addressOrAddressMap === 'string') address = addressOrAddressMap
-    else address = addressOrAddressMap[chainId]
-    if (!address) return null
+    if (!addressOrAddressMap || !ABI || !provider || !chainId) {
+      // Loading, waiting for chain or some basic information to instantiate the contract
+      return {
+        contract: null,
+        error: null,
+        chainId,
```

Comment:
`UseContractResult` says that `chainId: SupportedChainId`.
But here it might be falsy.